### PR TITLE
CloudStack modules: rename _facts -> _info

### LINKED ISSUE #53916: CloudStack task to list zones
##### SUMMARY
The CloudStack API provides a method `listZones` which is not available as in the `cs` module.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
Either implement this in `cs_facts` task and list cluster's zones in the facts, or create a new `cs_listzones` task.

##### ADDITIONAL INFORMATION
```yaml
- name: Get cluster zones
  cs_listzones:
  register: _cs_zones
```

Sample `_cs_zones`:
```
    [
        {
        "allocationstate": "Enabled",
        "dhcpprovider": "VirtualRouter",
        "id": "zone-id",
        "localstorageenabled": true,
        "name": "ch-gva-2",
        "networktype": "Basic",
        "securitygroupsenabled": true,
        "tags": [],
        "zonetoken": "zone-token"
        },
        {
        "allocationstate": "Enabled",
        "dhcpprovider": "VirtualRouter",
        "id": "zone-id",
        "localstorageenabled": true,
        "name": "ch-dk-2",
        "networktype": "Basic",
        "securitygroupsenabled": true,
        "tags": [],
        "zonetoken": "zone-token"
        },
        {
        "allocationstate": "Enabled",
        "dhcpprovider": "VirtualRouter",
        "id": "zone-id",
        "localstorageenabled": true,
        "name": "at-vie-1",
        "networktype": "Basic",
        "securitygroupsenabled": true,
        "tags": [],
        "zonetoken": "zone-token"
        },
        {
        "allocationstate": "Enabled",
        "dhcpprovider": "VirtualRouter",
        "id": "zone-id",
        "localstorageenabled": true,
        "name": "de-fra-1",
        "networktype": "Basic",
        "securitygroupsenabled": true,
        "tags": [],
        "zonetoken": "zone-token"
        }
    ]
```

### LINKED ISSUE #60515: CloudStack cloud _facts modules need to be renamed to _info and stop returning ansible_facts
##### SUMMARY
According to the [requirements](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_checklist.html#contributing-to-ansible-objective-requirements), `_facts` modules should return information that is specific to the host the module runs on as `ansible_facts`. (See also ansible/ansible#54280 for some discussion.)

I think that the following modules do not satisfy this condition: they return information that is not specific to the host machine, but bound to a cloud account. Please verify whether this is actually the case.

To fix this, the core team thinks the best approach is to rename the `_facts` modules to `_info` modules, and change the return value so that if the module is called with the `_facts` name (should be deprecated, and will then be removed in Ansible 2.13), it should return `ansible_facts` as now (and print a deprecation message), and if it is called with the `_info` name, it no longer returns `ansible_facts`, so that the result needs to be registered. This informs users that they have to change something, won't break their playbooks, and allows them (by switching to the `_info` names) to verify that their roles and playbooks no longer use the facts.

If that approach doesn't work, or you want to change the module's interface (like rename the return value(s)), it is better to deprecate the `_facts` module and create a new `_info` module next to it. To see both approaches implemented, you can take a look at https://github.com/ansible/ansible/commit/2d98734ad51c3de557c3a7d1f7166bdd114ceb0f (ansible/ansible#60178).

If you have any input — preferred method to solve this, doubts that all modules should be renamed, ... — please write it here!

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
~~cs_facts~~
cs_instance_facts
cs_zone_facts

##### ANSIBLE VERSION
```paste below
2.9.0
```
Please raise issues via the [new interface](https://github.com/ansible/ansible/issues/new/choose)

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Fixes #60515, fixes #53916.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
cs_instance_facts
cs_zone_facts

### CODE REVIEW

**Comment by resmo** [comment]
I though about changing the modules to return lists: list of zones, list of instances, with the possibility to filter by name to only get one item (but still a list). @dpassante any thoughts?

File: lib/ansible/modules/cloud/cloudstack/cs_instance_info.py [hunk 1]
```diff
@@ -43,17 +46,28 @@
...
+- name: Show the returned results of the registered variable
+  debug:
     var: vm
+
+# When the module is called as cs_instance_info, return values are also
+# published in ansible_facts['cloudstack_instance'] and can be used as
+# follows. Note that this is deprecated and will stop working in
+# Ansible 2.13.
+- name: gather instance info
+  cs_instance_facts:
```
Comment (samccann) [inline lib/ansible/modules/cloud/cloudstack/cs_instance_info.py]: I'm confused here - line 59 says 'when the module is called as cs_instance_info..." but on this example, it's using cs_instance_facts:

File: lib/ansible/modules/cloud/cloudstack/cs_zone_info.py [hunk 2]
```diff
@@ -39,6 +42,15 @@
   debug:
     var: zone
 
+# When the module is called as cs_zone_info, return values are also
+# published in ansible_facts['cloudstack_zone'] and can be used as
+# follows. Note that this is deprecated and will stop working in
+# Ansible 2.13.
+- name: Gather info from a zone
+  cs_zone_facts:
```
Comment (samccann) [inline lib/ansible/modules/cloud/cloudstack/cs_zone_info.py]: ditto

File: lib/ansible/modules/cloud/cloudstack/cs_instance_info.py (hunk 1 above)
Comment (felixfontein) [reply lib/ansible/modules/cloud/cloudstack/cs_instance_info.py]: Thanks! I must have copy'n'pasted that wrongly... Fixed both.

**Review by samccann — APPROVED** [review]
docs portion LGTM

File: lib/ansible/modules/cloud/cloudstack/cs_instance_info.py [hunk 3]
```diff
@@ -43,17 +46,28 @@
...
   delegate_to: localhost
   register: vm
 
-- debug:
-    var: cloudstack_instance
-
-- debug:
+- name: Show the returned results of the registered variable
+  debug:
     var: vm
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/cloudstack/cs_instance_info.py]: I'm a broken record, sorry
```suggestion
     msg: "{{ vm }}"
```
Could be done below and for cs_zone_info too.

File: lib/ansible/modules/cloud/cloudstack/cs_instance_info.py [hunk 4]
```diff
@@ -274,7 +288,7 @@ def __init__(self, module):
             'keypair': 'ssh_key',
             'hostname': 'host',
         }
-        self.facts = {
+        self.info = {
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/cloudstack/cs_instance_info.py]: It's weird, I can't find anything using self.facts. I think/hope this was just an unused thing.

File: lib/ansible/modules/cloud/cloudstack/cs_instance_info.py (hunk 3 above)
Comment (felixfontein) [reply lib/ansible/modules/cloud/cloudstack/cs_instance_info.py]: Thanks, changed it.

File: lib/ansible/modules/cloud/cloudstack/cs_instance_info.py (hunk 4 above)
Comment (felixfontein) [reply lib/ansible/modules/cloud/cloudstack/cs_instance_info.py]: Me neither. I removed it...

**Comment by dpassante** [comment]
@resmo I think it would be an interesting feature. However, returning a list for filtered items may break existing deployments that expect a dictionary.
These impacts could be mitigated by keeping the cloudstack_zone / cloudstack_instance exposure in the ansible_facts when there is only one item.

I also thought about writing a meta module that exposes lists of resources with the ability to filter by type of resources and by name to not break the existing but it seemed redundant with other modules and time-consuming to write and to maintain.

**Comment by felixfontein** [comment]
(Rebased to remove conflict.)

**Comment by felixfontein** [comment]
@dpassante if the return values should be modified, the old module `_facts` modules should be deprecated and new `_info` modules added (with the new return values). Existing playbooks using the `_facts` version should continue to run (with deprecation warning, and until `_facts` is gone).

**Comment by dpassante** [comment]
@felixfontein Indeed. So, it's a good time to change the values returned by the new _info module!

**Comment by felixfontein** [comment]
@dpassante @resmo PTAL!

**Comment by felixfontein** [comment]
CS tests run through. Sanity tests should be fixed with the last commits.

ready_for_review

File: lib/ansible/modules/cloud/cloudstack/_cs_instance_facts.py [hunk 5]
```diff
@@ -19,6 +19,10 @@
 short_description: Gathering facts from the API of instances from Apache CloudStack based clouds.
 description:
     - Gathering facts from the API of an instance.
+deprecated:
+  removed_in: "2.12"
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/cloudstack/_cs_instance_facts.py]: this should be 2.13, right?

File: lib/ansible/modules/cloud/cloudstack/_cs_zone_facts.py [hunk 6]
```diff
@@ -16,6 +16,10 @@
 description:
   - Gathering facts from the API of a zone.
   - Sets Ansible facts accessable by the key C(cloudstack_zone) and since version 2.6 also returns results.
+deprecated:
+  removed_in: "2.12"
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/cloudstack/_cs_zone_facts.py]: Here too

File: lib/ansible/modules/cloud/cloudstack/_cs_instance_facts.py (hunk 5 above)
Comment (felixfontein) [reply lib/ansible/modules/cloud/cloudstack/_cs_instance_facts.py]: Right! I guess I copied it from a facts→info rename for 2.8... Fixed; will push changes once CI ran through.

File: lib/ansible/modules/cloud/cloudstack/cs_zone_info.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+    aliases: [ name ]
+extends_documentation_fragment: cloudstack
+'''
+
+EXAMPLES = '''
+- name: Gather information from a zone
+  cs_zone_info:
+    zone: ch-gva-1
+  register: zone
+  delegate_to: localhost
```
Comment (dpassante) [inline lib/ansible/modules/cloud/cloudstack/cs_zone_info.py:38]: Perhaps we could add an example to show how to list zones (without filtering).

File: lib/ansible/modules/cloud/cloudstack/cs_instance_info.py [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+    type: str
+extends_documentation_fragment: cloudstack
+'''
+
+EXAMPLES = '''
+- name: gather instance information
+  cs_instance_info:
+    name: web-vm-1
+  delegate_to: localhost
+  register: vm
```
Comment (dpassante) [inline lib/ansible/modules/cloud/cloudstack/cs_instance_info.py:51]: Perhaps we could add an example to show how to list instances (without filtering).

**Review by dpassante — APPROVED** [review]
LGTM

**Comment by resmo** [comment]
shipit

**Comment by resmo** [comment]
Fixes #53916

File: lib/ansible/modules/cloud/cloudstack/cs_zone_info.py (hunk 7 above)
Comment (felixfontein) [reply lib/ansible/modules/cloud/cloudstack/cs_zone_info.py:38]: Added one, will push once tests ran through.

File: lib/ansible/modules/cloud/cloudstack/cs_instance_info.py (hunk 8 above)
Comment (felixfontein) [reply lib/ansible/modules/cloud/cloudstack/cs_instance_info.py:51]: Same here.

**Comment by felixfontein** [comment]
Final commits. Fix version number for deprecation and add examples.

**Comment by resmo** [comment]
shipit

**Comment by dpassante** [comment]
Thanks @felixfontein for these changes.

**Comment by dpassante** [comment]
shipit

**Comment by felixfontein** [comment]
@samccann @s-hertel @dpassante @resmo thank you all for helping getting this done! :)

**Pull request merged**